---
title: "ComplexHeatMap Test Script"
author: "Trevor Randall"
date: "12/05/2020"
output: html_document
---

Analysis from: https://jokergoo.github.io/ComplexHeatmap-reference/book/a-single-heatmap.html

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


#Clear Global Environment
```{r}
remove(list = ls())
```


```{r}
#Set Library Directory
#PC Path
.libPaths(c("L:/RStudios/RPackRatLibLocations", "L:/RStudios/RPackRat_2019_04_DESEQLibs"))

#Set working directory
#PC Path
setwd("E:/Dropbox/Dropbox/Harrison Lab - Trevor Randall/RNASeq Analysis/RNASeqAnlyPkrat_2020_03/CF39vsCF39S/Pure Temperature Analysis Comparisons/Individual Heatmaps/Heat Maps")
#Laptop Path
#Set Library Directory
#.libPaths(c(""))
#setwd("")
#sink(file = "./RSessionRawRun.txt") 
``` 


#Libraries required
```{r}
library(readr)
library(ComplexHeatmap)
library(circlize)
library(cluster)
library(dendextend)
#library(seriation)
```

PTA = Pure Temperature Analysis
#Dataset
```{r}
PAO1_Tags <- read_csv("../../../../LocusTags and PAO1 Loci ID's.csv")

LoadedFlagellarData <- read_delim("E:/Dropbox/Dropbox/Harrison Lab - Trevor Randall/RNASeq Analysis/RNASeqAnlyPkrat_2020_03/CF39vsCF39S/Pure Temperature Analysis Comparisons/Individual Heatmaps/Cut Data Sets/Flagella Genes Full Dataset.txt", "\t", escape_double = FALSE, trim_ws = TRUE)

DataSet <- LoadedFlagellarData

DataSet <-merge(DataSet, y = PAO1_Tags, by = "locus")
ColToRemove <- c("locusNumber.x", "locusNumber.y")
DataSet <- DataSet[, !names(DataSet) %in% ColToRemove, drop = F]

FinalColumnNames <- append(colnames(DataSet)[1:7], c("NucleotidePAO1Locus", "NucleotidePAO1ID", "ProteinPAO1Locus", "ProteinPAO1ID"))

colnames(DataSet) <- FinalColumnNames

```

<!-- Names of Columns = "FoldChange_25C", "FoldChange_29C", "FoldChange_33C", "FoldChange_37C", "FoldChange_41C" -->
<!-- ```{r} -->
<!-- DataSet$FoldChange_25C <- NULL -->
<!-- DataSet$FoldChange_29C <- NULL -->
<!-- ``` -->


```{r}
FoldChangeData <- data.frame(DataSet$locus, DataSet$FoldChange_25, DataSet$FoldChange_29, DataSet$FoldChange_33, DataSet$FoldChange_37, DataSet$FoldChange_41)

NamesOfColumns <- c("locus", "FoldChange_25", "FoldChange_29", "FoldChange_33", "FoldChange_37", "FoldChange_41")
colnames(FoldChangeData) <- NamesOfColumns
rownames(FoldChangeData) <- FoldChangeData$locus
FoldChangeData <- FoldChangeData[,-1]
```


```{r}
FoldChangeDataNoNA <- FoldChangeData
FoldChangeDataNoNA[is.na(FoldChangeDataNoNA)] <- 0
FoldChangeDataNoNAMat <- data.matrix(FoldChangeDataNoNA)
```

```{r}
mat <- FoldChangeDataNoNAMat
colorSetRedBlue <- colorRamp2(c(-10, 0, 10), c(rgb(0, 114/255, 178/255), "white", rgb(213/255, 94/255, 0))) #First color is Blue and second is Vermillion from the Nature Colorblind palette
ColOrder = order(as.numeric(gsub("FoldChange_", "", colnames(mat))))
```

#Coloration of row dendrigram
```{r}
row_dend = as.dendrogram(hclust(dist(mat, method = "maximum")), method = "complete")
row_dend = color_branches(row_dend, k = 2) # `color_branches()` returns a dendrogram object

split = data.frame(cutree(hclust(dist(mat, method = "maximum")), h=2))

```

```{r}
InCellTextSize = 7
RowLabelTextCell = 7
ColumnLabelText = 8
LegendName = "    Gene \nExpression\n"



HeatmapTest <- Heatmap(
        matrix = mat, 
        # cluster_rows = color_branches(as.dendrogram(hclust(dist(mat, method = "canberra")), method = "complete"), k = 2),
        # row_split = 3,
        # row_gap = unit(1.2, 'mm'),
        border = FALSE, 
        
        row_order = DataSet$locus,       
        row_dend_reorder = FALSE,
        show_row_names = FALSE,
        
        right_annotation = rowAnnotation(
        Col1 = anno_text(DataSet$locus, location = unit(0.75, "mm"), gp = gpar(fontsize = RowLabelTextCell)), 
        #Col4 = anno_text(DataSet$NucleotidePAO1Locus, location = unit(1, "mm"), gp = gpar(fontsize = RowLabelTextCell)),
      	Col3 = anno_text(DataSet$ProteinPAO1Locus, location = unit(0, "mm"), gp = gpar(fontsize = RowLabelTextCell)),
        #Col2 = anno_text(DataSet$gene, location = unit(0, "mm"), gp = gpar(fontsize = RowLabelTextCell)),
       # Col5 = anno_text(DataSet$NucleotidePAO1ID, location = unit(0, "mm"), gp = gpar(fontsize = RowLabelTextCell)),
        gap = unit(6, "mm")),
        
        cell_fun = function(j, i, x, y, width, height, fill) {
          if(mat[i, j] < 0 || mat[i, j] > 0)
            grid.text(sprintf("%.1f", mat[i, j]), x, y, gp = gpar(fontsize = InCellTextSize))},
        
        col = colorSetRedBlue,
        
        column_dend_reorder = FALSE, 
        column_order = ColOrder, 
        row_title = "Transcribed Gene Loci", 
        column_title = "Conditions tested (CF39S vs CF39)", 
        column_title_side = "bottom",
        show_row_dend = FALSE,
        #row_dend_width = unit(6, "cm"),
        row_names_gp = gpar(fontsize = RowLabelTextCell),
        
        heatmap_legend_param = list(title = LegendName)
        
        , raster_quality = 10
        )


HeatmapTest
```




```{r}
pdf("heatmap_Canberra(mat)AsDend_colored_split_CellLabled_AddCol_NoDend_MinProt.pdf", width = 10, height = 22)
HeatmapTest
dev.off()
```



```

#Session Info
```{r}
sink(file = "./SessionInfo.txt")
sessionInfo()
sink(file = NULL)
```


#References

The citation function can be used to  who you should be citing.

```{r}
sink(file = "./ReferenceInfo.txt")
citation("readr")
citation("ComplexHeatmap")
citation("circlize")
citation("cluster")
sink(file = NULL)
```
